src/neo/SmartContract/ApplicationEngine.Contract.cs

```diff
@@ -71,14 +72,14 @@ partial class ApplicationEngine
         /// <param name="args">The arguments to be used.</param>
         protected internal void CallContract(UInt160 contractHash, string method, CallFlags callFlags, Array args)
         {
-            if (method.StartsWith('_')) throw new ArgumentException($"Invalid Method Name: {method}");
+            if (method.StartsWith('_')) throw new CatchableException($"Invalid method name: {method}");
             if ((callFlags & ~CallFlags.All) != 0)
-                throw new ArgumentOutOfRangeException(nameof(callFlags));
+                throw new CatchableException($"Argument out of range: {nameof(callFlags)}");
 
             ContractState contract = NativeContract.ContractManagement.GetContract(Snapshot, contractHash);
-            if (contract is null) throw new InvalidOperationException($"Called Contract Does Not Exist: {contractHash}");
+            if (contract is null) throw new CatchableException($"Contract doesn't exist: {contractHash}");
             ContractMethodDescriptor md = contract.Manifest.Abi.GetMethod(method, args.Count);
-            if (md is null) throw new InvalidOperationException($"Method \"{method}\" with {args.Count} parameter(s) doesn't exist in the contract {contractHash}.");
+            if (md is null) throw new CatchableException($"Method \"{method}\" with {args.Count} parameter(s) doesn't exist in the contract {contractHash}.");
```

Comment:
`Token.transfer(A, B...)` makes `Token` probe `B` address in the `ContractManagement` and call `B.onNEP17Payment()`. If there is no `onNEP17Payment` method there the transaction would fail before the change, but now a caller will be able to catch an exception, which means he can persist any storage changes done by `Token` before `B.onNEP17Payment()` call even though they must not be persisted as per NEP-17.

Reply:
That's why I created https://github.com/neo-project/neo-devpack-dotnet/pull/735

Reply:
That's exactly

> protecting against this will complicate contract development

Especially given that C# is not the only language on the Neo platform.

Reply:
It's not c# only. Every language can implement this.

Reply:
There's nothing that forces developers to abort on exception in their onNEP17Payment/onNEP11Payment methods, even with we introduce a new mechanism in the C# compiler to make it easy for developers to do it. 

I don' think any of the exceptions in CallContract should be catchable